test/clj/rems/api/test_api.clj

```diff
@@ -26,23 +27,77 @@
+            csrf (get-csrf-token cookie)
+            resp (-> (request :post "/api/email/send-reminders")
+                     (header "Cookie" cookie)
+                     (header "x-csrf-token" csrf)
+                     (header "x-rems-api-key" "WRONG")
+                     handler)]
+        (is (response-is-forbidden? resp)))))
+  (testing "api key user whitelist"
+    (api-key/add-api-key! "43" {:comment "all users" :users nil})
+    (api-key/add-api-key! "44" {:comment "only alice & malice" :users ["alice" "malice"]})
+    (testing "> api key without whitelist can impersonate any user >"
+      (doseq [user ["owner" "alice" "malice"]]
+        (testing user
+          (is (= 200 (:status (api-response :get "/api/catalogue/" nil
+                                            "43" user)))))))
+    (testing "> api key with whitelist can only impersonate given users >"
+      (doseq [user ["alice" "malice"]]
+        (testing user
+          (is (= 200 (:status (api-response :get "/api/catalogue/" nil
+                                            "44" user))))))
+      (is (response-is-unauthorized? (api-response :get "/api/catalogue/" nil
+                                                   "44" "owner")))))
+  (testing "api key path whitelist"
+    (api-key/add-api-key! "45" {:comment "all paths" :paths nil})
+    (api-key/add-api-key! "46" {:comment "limited paths" :paths [{:method "any"
+                                                                  :path "/api/applications"}
+                                                                 {:method "any"
+                                                                  :path "/api/my-applications"}]})
+    (api-key/add-api-key! "47" {:comment "regex path" :paths [{:method "any"
+                                                               :path "/api/c.*"}
+                                                              {:method "get"
+                                                               :path "/api/users/.*"}]})
+    (testing "> api key without whitelist can access any path >"
+      (doseq [path ["/api/applications" "/api/my-applications" "/api/catalogue"]]
+        (testing path
+          (is (= 200 (:status (api-response :get path nil
+                                            "45" "owner")))))))
+    (testing "> api key with whitelist can access only given paths >"
+      (doseq [path ["/api/applications" "/api/my-applications"]]
+        (testing path
+          (is (= 200 (:status (api-response :get path nil
+                                            "46" "owner"))))))
+      (is (response-is-unauthorized? (api-response :get "/api/catalogue" nil
+                                                   "46" "owner"))))
+    (testing "> api key with whitelist can access only matching paths >"
+      (doseq [path ["/api/catalogue?query=param" "/api/catalogue-items"]]
+        (testing path
+          (is (= 200 (:status (api-response :get path nil
```

Comment:
Maybe see if writing a local helper fn would clean the test cases?

Reply:
added `response-is-ok?`